### ISSUE TITLE
Regression in v1.1.1: spawn makes system unstable and silently crashing afterwards

### ISSUE THREAD

**Issue by mzoliker** [ish-app/ish#1244]
Hi, I have been using iSH for a while now (thanks!) and v1.1.1 made my code crash. I took the time to reinstall everything on a fresh iSH install and now I can confirm I have found a regression in v1.1.1, that I could track down to the use of the spawn function, which makes the system unstable (no error in thrown) but it crashes silently a few instructions afterwards with something like an infinite loop (but CTRL-C or CTRL-Z do nothing, I have to close iSh).

More context:
- This is a nodejs program
- Which is using the native module node-sqlite3 (running the binary compiled with the previous version of iSH, as the v1.1.1 is not able to compile it any more)
- Which is using the npm module 'node-pre-gyp' to startup the module
- Which is using the npm module 'detect-libc' to detect musl or glibc
- The module 'detect-libc' uses two spawn calls that make the system unstable (https://github.com/lovell/detect-libc/blob/master/lib/detect-libc.js). This was not the case with previous version of iSH, as I had no problem with the exact same code

Why is it important?
- It breaks 'node-pre-gyp' (which uses spawn to run python), so all nodejs native modules cannot be compiled any more
- Even if you already have the compiled binary (my case) it still does not work, as 'detect-libc' is executed at run time, so it requires to patch dependencies not to use spawn to get it to work as it did before

Using iSH v1.1.1 on iPhone XS Max iOS v14.2. I hope this is clear enough but feel free to ask anything if I can help you to solve this issue!

Kind regards,
Maurice

**Comment by tbodt** [comment]
Thanks for the report! Could you provide instructions or a script (ideally as simple as possible) to reproduce the crashing?

**Comment by mzoliker** [comment]
Yes, of course, here is a very simple nodejs script that crashes, just put it into a file named "test.js":
const spawnSync = require('child_process').spawnSync;
const res = spawnSync('ls', [], { encoding: 'utf8' });
console.log(res); //it "works"
const nb = 10000;
for (let i = 1; i <= nb; i++) { //... but crashes before reaching 10000
  console.log(`${i}/${nb}`);
}
//EOF

And then in iSH:
apk add nodejs
node test.js

Hope this helps! :-)

Kind regards,
Maurice

PS: spawnSync function documentation: https://nodejs.org/docs/latest-v12.x/api/child_process.html#child_process_child_process_spawnsync_command_args_options

**Comment by emkey1** [comment]
I believe the changes in this diff will greatly improve this, assuming it is the race condition I'm thinking of.

+++ b/kernel/exit.c
@@ -50,8 +50,10 @@ noreturn void do_exit(int status) {
     }
 
     // release all our resources
+    lock(&current->general_lock); 
     mm_release(current->mm);
     current->mm = NULL;
+    unlock(&current->general_lock); 
     fdtable_release(current->files);
     current->files = NULL;
     fs_info_release(current->fs);
diff --git a/kernel/fork.c b/kernel/fork.c
index 3072053..49f0016 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -132,7 +132,9 @@ fail_free_fs:
 fail_free_files:
     fdtable_release(task->files);
 fail_free_mem:
+    lock(&current->general_lock); 
     mm_release(task->mm);
+    unlock(&current->general_lock); 
     return err;
 }

**Comment by thisIsP0G11** [comment]
Some fùčkíńg 8============D sêxër just closed my issue >:( >:(

**Comment by mzoliker** [comment]
Hi!
Thanks emkey1 for your patch!
I think we will need to wait for tbodt’s feedback to ensure it works with the test script I provided :-).
Thank you for your help!
Kind regards,
Maurice

**Comment by emkey1** [comment]
I just tried the test script on a build of iSH that includes the above patch.  The patch might help, but it does not fix the issue unfortunately.  iSH still becomes unstable and crashes soon after the script is run.

**Comment by mzoliker** [comment]
Hi,

Any news on this issue?

As it is a regression from previous version, would it be possible to review changes since last version that could have such an impact?

I really like iSH and I am a bit disappointed to see it crash where it worked fine before.

Thanks for your help, and feel free to ask if I can do something on my side to help you solve this.

Kind regards,
Maurice
